firewalldb/sql_migration.go

```diff
@@ -0,0 +1,478 @@
+// bbolt database to a SQL database. The migration is done in a single
+// transaction to ensure that all rows in the stores are migrated or none at
+// all.
+//
+// Note that this migration currently only migrates the kvstores, but will be
+// extended in the future to also migrate the privacy mapper and action stores.
+//
+// NOTE: As sessions may contain linked sessions and accounts, the sessions and
+// accounts sql migration MUST be run prior to this migration.
+func MigrateFirewallDBToSQL(ctx context.Context, kvStore *BoltDB,
+	tx SQLQueries) error {
+
+	log.Infof("Starting migration of the rules DB to SQL")
+
+	err := migrateKVStoresDBToSQL(ctx, kvStore, tx)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("The rules DB has been migrated from KV to SQL.")
+
+	// TODO(viktor): Add migration for the privacy mapper and the action
+	// stores.
+
+	return nil
+}
+
+// migrateKVStoresDBToSQL runs the migration of all KV stores from the KV
+// database to the SQL database. The function also asserts that the
+// migrated values match the original values in the KV store.
+// See the illustration in the firwalldb/kvstores_kvdb.go file to understand
+// the structure of the KV stores, and why we process the buckets in the
+// order we do.
+// Note that this function and the subsequent functions are intentionally
+// designed to loop over all buckets and values that exist in the KV store,
+// so that we are sure that we actually find all stores and values that
+// exist in the KV store, and can be sure that the kv store actually follows
+// the expected structure.
+func migrateKVStoresDBToSQL(ctx context.Context, kvStore *BoltDB,
+	sqlTx SQLQueries) error {
+
+	log.Infof("Starting migration of the KV stores to SQL")
+
+	// allParams will hold all the kvParams that are inserted into the
+	// SQL database during the migration.
+	var allParams []kvParams
+
+	err := kvStore.Update(func(kvTx *bbolt.Tx) error {
+		for _, perm := range []bool{true, false} {
+			mainBucket, err := getMainBucket(kvTx, true, perm)
+			if err != nil {
+				return err
+			}
+
+			err = mainBucket.ForEach(func(k, v []byte) error {
+				if v != nil {
+					return errors.New("expected only " +
+						"buckets under main bucket")
+				}
+
+				ruleName := k
+				ruleNameBucket := mainBucket.Bucket(k)
+				if ruleNameBucket == nil {
+					return fmt.Errorf("rule bucket %s "+
+						"not found", string(k))
+				}
+
+				ruleId, err := sqlTx.GetOrInsertRuleID(
+					ctx, string(ruleName),
+				)
+				if err != nil {
+					return err
+				}
+
+				params, err := processRuleBucket(
+					ctx, sqlTx, perm, ruleId,
+					ruleNameBucket,
+				)
+				if err != nil {
+					return err
+				}
+
+				allParams = append(allParams, params...)
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// After the migration is done, we validate that all inserted kvParams
```

Comment:
cool yeah thanks for separating this 👍  i think it is especially important for the kv stores since there is a higher chance of overwriting 👍